types-tests/message.test-d.ts

```diff
@@ -70,3 +68,29 @@ expectType<void>(
     await Promise.resolve(message);
   }),
 );
+
+// Resolve the event type when having subtype in a listener constraint
+// app.message<'message_replied'>('foo', async ({ message }) => {});
+
+expectType<void>(
+  app.message<'thread_broadcast'>('foo', async ({ message }) => {
+    expectNotType<MessageEvent>(message);
+    expectType<ThreadBroadcastMessageEvent>(message);
+    message.channel; // the property access should compile
+    message.thread_ts; // the property access should compile
+    message.ts; // the property access should compile
+    message.root; // the property access should compile
+    await Promise.resolve(message);
+  }),
+);
+expectType<void>(
+  // no subtype in the event payload
+  app.message<undefined>('foo', async ({ message }) => {
```

Comment:
Setting undefined as the default can be a breaking changes to the existing apps.